x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx

```diff
@@ -77,14 +77,17 @@ export class MachineLearningFlyout extends Component<FlyoutProps, FlyoutState> {
     this.setState({ isLoading: true });
     try {
       const { serviceName, transactionType } = this.props.urlParams;
-      if (serviceName) {
-        const res = await startMLJob({ serviceName, transactionType });
-        const didSucceed = res.datafeeds[0].success && res.jobs[0].success;
-        if (!didSucceed) {
-          throw new Error('Creating ML job failed');
-        }
-        this.addSuccessToast();
+      if (!serviceName || !transactionType) {
+        throw new Error(
+          'Service name and transaction type are required to create this ML job'
+        );
```

Comment:
I'm wondering if we should throw here. Previously we'd just `return` if `serviceName` was `undefined`.
When would this happen?

Reply:
https://github.com/elastic/kibana/pull/30284#commitcomment-32218351

Reply:
But it can never happen, right? At least not if you add `transactionType` in `render`:
https://github.com/elastic/kibana/blob/c96e83d149eb1077f1cfe93da023f6e17f31bedd/x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx#L171-L173

Reply:
Typescript obviously doesn't know this since all the `urlParams` are optional. Since we only use `serviceName` and `transactionType` from `urlParams` perhaps we should specifically pass those in, and make them required on the `MachineLearningFlyout` interface. That way we can statically make sure that they will always be available, and won't need checks and throws in runtime.

Reply:
I know this is outside the scope of this PR, just feels like a nice TS change

Reply:
Right, without those guarantees as it is now, I think it's better to throw so that we produce the right error toast to the user. Looking at the code, there's no way to know for sure that it definitely can't happen, at least not how it's written now.

I like the idea of moving toward passing individual props to clean this kind of thing up.